src/tensor_ops/boolean/mod.rs

```diff
@@ -0,0 +1,320 @@
+mod cpu_kernels;
+
+#[cfg(feature = "cuda")]
+mod cuda_kernels;
+
+use crate::{
+    prelude::{OnesTensor, Tensor, ZerosTensor},
+    shapes::*,
+    tensor::DeviceStorage,
+};
+
+use std::ops::{BitAnd, BitOr, BitXor, Not};
+
+use super::Device;
+
+pub trait BooleanKernel: DeviceStorage + OnesTensor<bool> + ZerosTensor<bool> {
+    fn not<S: Shape>(
+        &self,
+        inp: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn and<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn or<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn xor<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn scalar_and<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: bool,
+    ) -> Result<Self::Storage<S, bool>, Self::Err> {
+        if rhs {
+            Ok(lhs.clone())
+        } else {
+            Ok(self.try_zeros_like(lhs)?.storage)
+        }
+    }
+
+    fn scalar_or<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: bool,
+    ) -> Result<Self::Storage<S, bool>, Self::Err> {
+        if rhs {
+            Ok(self.try_ones_like(lhs)?.storage)
+        } else {
+            Ok(lhs.clone())
+        }
+    }
+
+    fn scalar_xor<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: bool,
+    ) -> Result<Self::Storage<S, bool>, Self::Err> {
+        if rhs {
+            self.not(lhs)
+        } else {
+            Ok(lhs.clone())
+        }
+    }
+}
+
+impl<S: Shape, D: BooleanKernel> Not for Tensor<S, bool, D> {
+    type Output = Self;
+
+    fn not(self) -> Self {
+        !&self
+    }
+}
+
+impl<S: Shape, D: BooleanKernel> Not for &Tensor<S, bool, D> {
+    type Output = Tensor<S, bool, D>;
+
+    fn not(self) -> Self::Output {
+        self.device.upgrade(self.device.not(&self.storage).unwrap())
+    }
+}
+
+macro_rules! boolean_op_impl {
+    ($op:ident, $method1:ident, $method2:ident, $method3:ident) => {
+        impl<S: Shape, D: BooleanKernel> $op for Tensor<S, bool, D> {
+            type Output = Self;
+
+            fn $method1(self, rhs: Self) -> Self {
+                self.device
+                    .upgrade(self.device.$method2(&self.storage, &rhs.storage).unwrap())
```

Comment:
Related to #420, but all of these need to assert that the shapes are equal for runtime shapes. `assert_eq!(self.shape(), rhs.shape()`